File: apps/yagye_portal/app/controllers/checkout/payment_links_controller.rb
# frozen_string_literal: true

module Checkout
  class PaymentLinksController < ApplicationController
    def index
      authorize :checkout, :index?
      result = core.list_payment_links(merchant_code: current_user.merchant_code)
      links  = result.success? ? result.body["data"] : []
      links  = filter_links(links)
      render Checkout::PaymentLinksIndexView.new(
        links:         links,
        query:         params[:q],
        view:          params[:view].presence_in(%w[list grid]) || "list",
        active_filter: params[:active]
      )
    end

    def filter
      authorize :checkout, :index?
      render Checkout::PaymentLinksFilterView.new(
        query:         params[:q],
        active_filter: params[:active],
        view:          params[:view]
      )
    end

    def show
      authorize :checkout, :index?
      result = core.get_payment_link(params[:id])
      return redirect_to payment_links_path, alert: "Payment link not found." unless result.success?
      render Checkout::PaymentLinks::ShowView.new(link: result.body)
    end

    def deactivate
      authorize :checkout, :create?
      result = core.deactivate_payment_link(params[:id])
      if result.success?
        redirect_to payment_link_path(params[:id]), notice: "Payment link deactivated."
      else
        redirect_to payment_link_path(params[:id]), alert: result.error_message
      end
    end

    def new
      authorize :checkout, :new?
      render Checkout::PaymentLinkFormView.new(mode: Current.mode)
    end

    def create
      authorize :checkout, :create?

      kind = params[:kind].presence || "fixed_amount"
      if kind == "fixed_amount" && params[:amount].blank?
        return render Checkout::PaymentLinkFormView.new(errors: ["Amount is required for fixed-amount links"], mode: Current.mode),
                      status: :unprocessable_entity
      end

      allowed_methods = Array(params[:allowed_methods]).reject(&:blank?)
      if allowed_methods.empty?
        return render Checkout::PaymentLinkFormView.new(errors: ["Select at least one accepted payment method"], mode: Current.mode),
                      status: :unprocessable_entity
      end

      reusable = params[:reusable] == "1"

      result = core.create_payment_link(
        merchant_code:   current_user.merchant_code,
        kind:            kind,
        currency:        params[:currency].presence || "GHS",
        description:     params[:description],
        amount:          parse_amount(params[:amount]),
        allowed_methods: allowed_methods,
        collect_email:   params[:collect_email] == "1",
        collect_phone:   params[:collect_phone] == "1",
        collect_name:    params[:collect_name] == "1",
        reusable:        reusable,
        max_uses:        (reusable && params[:max_uses].present?) ? params[:max_uses].to_i : nil,
        expires_at:      params[:expires_at].presence
      )

      if result.success?
        redirect_to payment_link_layout_path(result.body["id"]),
                    notice: "Payment link created. Configure the checkout layout below."
      else
        render Checkout::PaymentLinkFormView.new(errors: [ result.error_message ], mode: Current.mode),
               status: :unprocessable_entity
      end
    end

    def layout
      authorize :checkout, :manage_layout?
      result = core.list_payment_links(merchant_code: current_user.merchant_code)
      @links = result.success? ? result.body["data"] : []
      @current = @links.find { |l| l["id"] == params[:id] }
      return redirect_to payment_links_path, alert: "Payment link not found." if @current.nil?

      render Checkout::CheckoutLayoutView.new(link: @current, all_links: @links)
    end

    def update_layout
      authorize :checkout, :manage_layout?
      raw = params[:layout]
      layout = raw.is_a?(String) ? JSON.parse(raw) : raw&.to_unsafe_h || {}
      result = core.update_payment_link_layout(
        params[:id],
        merchant_code: current_user.merchant_code,
        layout:        layout
      )

      if result.success?
        render json: { ok: true, checkout_url: result.body["checkout_url"] }
      else
        render json: { ok: false, error: result.error_message }, status: :unprocessable_entity
      end
    end

    private

    def filter_links(links)
      links = links.reject { |l| l["kind"] == "invoice" }
      links = links.select { |l| l["description"].to_s.downcase.include?(params[:q].downcase) ||
                                 l["url_slug"].to_s.downcase.include?(params[:q].downcase) } if params[:q].present?
      if params[:active].present?
        active = params[:active] == "true"
        links = links.select { |l| l["active"] == active }
      end
      links
    end

    def parse_amount(raw)
      return nil if raw.blank?
      (raw.to_f * 100).round
    end

    def core
      @core ||= CoreApiClient.new
    end
  end
end

File: apps/yagye_core/lib/yagye_core_web/controllers/internal/payment_links_controller.ex
defmodule YagyeCoreWeb.Controllers.Internal.PaymentLinksController do
  @moduledoc false

  use YagyeCoreWeb, :controller

  alias YagyeCore.Merchants
  alias YagyeCore.PaymentLinks
  alias YagyeCoreWeb.Controllers.PaymentLinks.PaymentLinkJSON

  action_fallback YagyeCoreWeb.FallbackController

  # PATCH /internal/payment-links/:public_id/checkout-layout
  # Body: { merchant_code, layout }
  def update_checkout_layout(conn, %{"public_id" => public_id} = params) do
    merchant_code = params["merchant_code"]
    layout = params["layout"] || %{}

    with {:ok, merchant} <- Merchants.get_merchant(merchant_code),
         {:ok, link} <- PaymentLinks.update_checkout_layout(public_id, merchant.id, layout) do
      conn
      |> put_status(:ok)
      |> json(PaymentLinkJSON.data(link))
    end
  end

  # GET /internal/payment-links?merchant_code=:code
  def index(conn, %{"merchant_code" => merchant_code} = params) do
    with {:ok, merchant} <- Merchants.get_merchant(merchant_code) do
      opts = cursor_opts(params)
      {:ok, page} = PaymentLinks.list_links(merchant.id, opts)

      conn
      |> put_status(:ok)
      |> json(PaymentLinkJSON.list(page))
    end
  end

  # POST /internal/payment-links
  # Body: { merchant_code, kind, currency, description, ... }
  @allowed_create_fields ~w[kind currency description amount reusable
                             collect_email collect_phone collect_name
                             image_url allowed_methods max_uses expires_at metadata]

  def create(conn, %{"merchant_code" => merchant_code} = params) do
    attrs =
      Map.take(params, @allowed_create_fields)
      |> Map.new(fn {k, v} -> {String.to_atom(k), v} end)

    with {:ok, merchant} <- Merchants.get_merchant(merchant_code),
         {:ok, link} <- PaymentLinks.create_link(merchant.id, attrs) do
      conn
      |> put_status(:created)
      |> json(PaymentLinkJSON.data(link))
    end
  end

  defp cursor_opts(params) do
    [
      limit: params["limit"],
      starting_after: params["starting_after"]
    ]
    |> Keyword.reject(fn {_, v} -> is_nil(v) end)
  end
end
